### PR TITLE
Implement daily upload limit

### PR DESCRIPTION
**Pull request by Nutomic**
My instance [Peertube.social](https://peertube.social) is running on a rather slow VPS, but it has a lot of storage space. That means I don't really care how much data someone uploads in total, but if someone uploads a few GB of video in one day, it can create a transcoding backlog on the server for the next days. So I have a rule that only allows 1 GB of uploads per day, but it gets ignored at times. Having this rule coded into Peertube would be much prefered. This would solve #652 for me (even though I didn't mention this approach in the issue).

This is the first time I've written Typescript or server/side Javascript, so the code is probably not great. Some potential issues:

- `videoQuota` is used in a ton of places, so I may have missed some places for `videoQuotaDaily`
- database migration is untested
- new install isn't tested (after deleting storage/ and the Postgres database, I still had the old config for some reason)
- Formatting for the "daily upload limit reached" error is not working (parameters are not substituted). But I have the same problem with the existing "upload limit reached" error, so I'm not sure if this was already broken.

Edit: I also didn't run any tests yet, but the basic functionality is there.

### CODE REVIEW

File: .gitignore [hunk 1]
```diff
@@ -40,3 +40,4 @@
 /*.asc
 /server/tools/import-mediacore.ts
 /docker-volume/
+/nbproject/private/
```
Comment (Chocobozzz) [inline .gitignore]: ? Please remove and add it in your global git ignore

File: server/models/account/user.ts [hunk 2]
```diff
@@ -336,11 +374,16 @@ export class UserModel extends Model<UserModel> {
   }
 
   isAbleToUploadVideo (videoFile: { size: number }) {
-    if (this.videoQuota === -1) return Promise.resolve(true)
+      if (this.videoQuota === -1 && this.videoQuotaDaily === -1)
```
Comment (Chocobozzz) [inline server/models/account/user.ts]: Keep the return on the same line than `if`

File: server/models/account/user.ts [hunk 3]
```diff
@@ -336,11 +374,16 @@ export class UserModel extends Model<UserModel> {
   }
 
   isAbleToUploadVideo (videoFile: { size: number }) {
-    if (this.videoQuota === -1) return Promise.resolve(true)
+      if (this.videoQuota === -1 && this.videoQuotaDaily === -1) 
+        return Promise.resolve(true)
 
     return UserModel.getOriginalVideoFileTotalFromUser(this)
```
Comment (Chocobozzz) [inline server/models/account/user.ts]:
```
const [ totalBytes, totalBytesDaily ] = await Promise.all([
  UserModel.getOriginalVideoFileTotalFromUser(this),
  UserModel.getOriginalVideoFileTotalDailyFromUser(this)
])
```

to do things in parallel

File: server/tests/api/check-params/config.ts [hunk 4]
```diff
@@ -48,7 +48,8 @@ describe('Test config API validators', function () {
       email: 'superadmin1@example.com'
     },
     user: {
-      videoQuota: 5242881
+      videoQuota: 5242881,
+      videoQuotaDaily: -1
```
Comment (Chocobozzz) [inline server/tests/api/check-params/config.ts]: Use another value

File: server/tests/api/server/config.ts [hunk 5]
```diff
@@ -152,7 +152,8 @@ describe('Test config', function () {
         email: 'superadmin1@example.com'
       },
       user: {
-        videoQuota: 5242881
+        videoQuota: 5242881,
+        videoQuotaDaily: -1
```
Comment (Chocobozzz) [inline server/tests/api/server/config.ts]: Use another value

File: server/tests/utils/server/config.ts [hunk 6]
```diff
@@ -80,7 +80,8 @@ function updateCustomSubConfig (url: string, token: string, newConfig: any) {
       email: 'superadmin1@example.com'
     },
     user: {
-      videoQuota: 5242881
+      videoQuota: 5242881,
+      videoQuotaDaily: -1
```
Comment (Chocobozzz) [inline server/tests/utils/server/config.ts]: Use another value

File: shared/models/users/user-video-quota.model.ts [hunk 7]
```diff
@@ -1,3 +1,4 @@
 export interface UserVideoQuota {
   videoQuotaUsed: number
-}
+  videoQuotaUsedDaily: number
+}
```
Comment (Chocobozzz) [inline shared/models/users/user-video-quota.model.ts]: Add a newline

File: server/models/account/user.ts (hunk 2 above)
Comment (Nutomic) [reply server/models/account/user.ts]: That makes a pretty long line, but okay.

File: server/models/account/user.ts (hunk 3 above)
Comment (Nutomic) [reply server/models/account/user.ts]: Ah I was wondering how to do that.

**Comment by Nutomic** [comment]
Fixed the stuff you mentioned. How can I tell why the tests failed? All I'm getting in the logs is this:
```
 Error: expected 200 "OK", got 500 "Internal Server Error"
  at Test._assertStatus (node_modules/supertest/lib/test.js:268:12)                                                                                                                     
  at Test._assertFunction (node_modules/supertest/lib/test.js:283:11)                                                                                                                   
  at Test.assert (node_modules/supertest/lib/test.js:173:18)                                                                                                                            
  at assert (node_modules/supertest/lib/test.js:131:12)                                                                                                                                 
  at /home/felix/workspace/peertube/node_modules/supertest/lib/test.js:128:5                                                                                                            
  at Test.Request.callback (node_modules/superagent/lib/node/index.js:718:3)                                                                                                            
  at IncomingMessage.parser (node_modules/superagent/lib/node/index.js:906:18)                                                                                                          
  at endReadableNT (_stream_readable.js:1085:12)                                                                                                                                        
  at process._tickCallback (internal/process/next_tick.js:63:19) 
```

**Comment by Chocobozzz** [comment]
@Nutomic Chec the logs in `test1/...`

**Comment by Nutomic** [comment]
Okay I fixed the tests, added a few new ones, and rebased on develop.

**Comment by Chocobozzz** [comment]
@Nutomic Nice! There are some `// TODO`: what do you want to do? We merge and see them after or do you want to implement them in this PR?

**Comment by Nutomic** [comment]
Oops I forgot about those, just wrote them during debugging and forgot to remove them. Fixed now!

**Comment by Nutomic** [comment]
Btw looks like something is wrong with the tests on develop?

**Comment by Chocobozzz** [comment]
> Btw looks like something is wrong with the tests on develop?

Yes, my bad :) Thanks for your PR @Nutomic!

**Pull request merged**